File: src/com/train/algorithm/list/Subsets.java
package com.train.algorithm.list;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

// 子集问题(回朔法)
public class Subsets {
    // leetcode-78
    List<Integer> temp = new ArrayList<Integer>();
    List<List<Integer>> res = new ArrayList<List<Integer>>();
    public List<List<Integer>> subsets(int[] nums) {
        dfs(0,nums);
        return res;
    }
    public void dfs(int cur,int[] nums){
        if (cur == nums.length) {
            res.add(new ArrayList<Integer>(temp));
            return;
        }
        temp.add(nums[cur]);
        dfs(cur+1,nums);
        temp.remove(temp.size() - 1);
        dfs(cur+1,nums);
    }

    // leetcode-46
    public List<List<Integer>> permute(int[] nums) {
        List<Integer> temp = new ArrayList<Integer>();
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        for(int num:nums){
            temp.add(num);
        }
        int n = nums.length;
        dfs(n,temp,res,0);
        return res;
    }
    public void dfs(int n,List<Integer> temp,List<List<Integer>> res,int first){
        if (first == n) {
            res.add(new ArrayList<Integer>(temp));
        }
        for (int i = first;i < n ;i++ ) {
            Collections.swap(temp,first,i);
            dfs(n,temp,res,first+1);
            Collections.swap(temp,first,i);
        }
    }

    // leetcode-784
    public List<String> letterCasePermutation(String S) {
        List<String> res = new ArrayList<>();
        char[] charArray = S.toCharArray();
        dfs(charArray,0,res);
        return res;
    }
    private void dfs(char[] charArray,int index,List<String> res){
        if (index == charArray.length) {
            res.add(new String(charArray));
            return;
        }
        dfs(charArray,index+1,res);
        if (Character.isLetter(charArray[index])) {
            // 转大写
            charArray[index] ^= 1 << 5;
            dfs(charArray,index+1,res);
        }
    }
}

File: src/com/train/algorithm/search/implementInC++/Subsets.cpp
//
// Created by 罗斯 on 7/9/2022.
//
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
//        vector<vector<int>> res;
//        vector<int> cur;
//        for (int i = 0; i <= nums.size(); ++i) {
//            dfs(nums,i,0,res,cur);
//        }
//        return res;
        const int n = nums.size();
        vector<vector<int>> res;
        for (int i = 0; i < 1 << n ; ++i) {
            vector<int> cur;
            for (int j = 0; j < n; ++j) {
                if (i & (1 << j)) cur.push_back(nums[j]);
            }
            res.push_back(cur);
        }
        return res;
    }
private:
    void dfs(vector<int>& nums,int n,int s,vector<vector<int>>& res,vector<int>& cur){
        if(n == cur.size()){
            res.push_back(cur);
            return;
        }

        for (int i = s; i < nums.size(); ++i) {
            cur.push_back(nums[i]);
            dfs(nums,n,i+1,res,cur);
            cur.pop_back();
        }
    }
};